feat: Move to ZipFolderPlugin for downloading multiple-nodes

susnux opened the pull request:
Fix https://github.com/nextcloud/server/issues/41037

## Summary

This contains multiple "things":
- :sparkle: Allow `GET` on folders to download them
  - if no accept header for ZIP is sent then (for browser download) allow GET params but if nothing set then fallback to Sabre -> should be non-breaking
- :bug: This now also allows downloading "select-multiple" files form e.g. "recent" or "favorites" view where the files are not really located in the same folder.



## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

susnux committed:
feat(dav): New `ZipFolderPlugin` which allows to download folders using GET requests

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix: Allow `Streamer` to specify type in constructor instead of magin UA handling

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
feat(files): Make the files download action use WebDAV zip download

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented on apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:
```diff
@@ -0,0 +1,133 @@
+		}
+	}
+
+	public function handleDownload(Request $request, Response $response) {
```
how to type `void|false`?

come-nc commented on apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:
```diff
@@ -0,0 +1,136 @@
+
+	public function handleDownload(Request $request, Response $response): null|false {
+		$node = $this->tree->getNodeForPath($request->getPath());
+		if (!($node instanceof ICollection)) {
+			// skip if not a collection
+			return null;
+		}
+
+		if (!($node instanceof \OCA\DAV\Connector\Sabre\Directory)) {
+			// only handle directories
+			return null;
+		}
```
```suggestion
		if (!($node instanceof \OCA\DAV\Connector\Sabre\Directory)) {
			// only handle directories
			return null;
		}
```
directory is always a collection?

come-nc commented on apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:
```diff
@@ -0,0 +1,136 @@
+		}
+
+		$query = $request->getQueryParameters();
+		$filesParam = $query['files'] ?? '';
+		$files = $request->getHeaderAsArray('X-NC-Files');
```
What happens if both are filled?

susnux replied:
It will be overwritten with the GET params as they need to take precedence to allow "plain browser GET".

susnux committed:
chore: compile assets

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

skjnldsv commented on apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:
```diff
@@ -0,0 +1,127 @@
+	 */
+	public function initialize(Server $server): void {
+		$this->server = $server;
+		$this->server->on('method:GET', $this->handleDownload(...), 100);
```
What is the `...` for? :thinking:

come-nc replied:
It is the same as `[$this, 'handleDownload']` but without typo risk. (Actually I think it’s the same as `Closure::fromCallable([$this, 'handleDownload'])` but you get the idea)

skjnldsv replied:
Arthur solved the issue, I'm smarter now :sparkles:

come-nc commented on apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:
```diff
@@ -0,0 +1,127 @@
+		}
+	}
+
+	public function handleDownload(Request $request, Response $response): null|false {
```
Do you really need to return `null|false`? What’s the meaning of the return type?
To support PHP 8.1 you’ll have to use `bool|null` and only type `false|null` in phpdoc.

susnux replied:
```suggestion
	/**
	 * @return false|null
	 */
	public function handleDownload(Request $request, Response $response): bool|null {
```

susnux commented on apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:
```diff
@@ -0,0 +1,127 @@
+			$content[] = $child->getNode();
+		}
+
+		$streamer = new Streamer($tarRequest, -1, count($content));
```
It is only used for checking if their is accept tar **OR** zip, if nothing of both then it just falls back to Sabre default behavior.
So in this place we only have either tar === true or/and zip === true. 
Meaning if tar is set we provide a tar ball, otherwise a zip archive

skjnldsv replied:
Right, just felt weird to have an unused variable :see_no_evil:

susnux committed:
feat(files_sharing): Make `ShareController` download route use the DAV `ZipFolderPlugin`

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
chore: Drop unused legacy `OC_Files`

It is replaced with the Sabre `ZipFolderPlugin` and apps should use the `OCP\AppFramework\Http\StreamResponse`.

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented on apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:
```diff
@@ -0,0 +1,136 @@
+		$accept = $request->getHeaderAsArray('Accept');
+		$acceptParam = $query['accept'] ?? '';
+		if ($acceptParam !== '') {
+			$accept = array_map(fn (string $name) => strtolower(trim($name)), explode(',', $acceptParam));
```
https://github.com/vimeo/psalm/issues/11075
